src/Mutator/MutatorFactory.php

```diff
@@ -0,0 +1,183 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutator;
+
+use function array_key_exists;
+use function class_exists;
+use Infection\Mutator\Util\Mutator;
+use Infection\Mutator\Util\MutatorConfig;
+use InvalidArgumentException;
+use function sprintf;
+use stdClass;
+
+/**
+ * @internal
+ */
+final class MutatorFactory
+{
+    /**
+     * @param array<string, bool|array<string, string>> $mutatorSettings
+     *
+     * @return array<string, Mutator>
+     */
+    public function create(array $mutatorSettings): array
+    {
+        return self::createFromNames(
+            self::retrieveMutatorNames($mutatorSettings)
+        );
+    }
+
+    /**
+     * @param array<string, bool|array<string, string>> $mutatorSettings
+     *
+     * @return array<string, array<string, string>>
+     */
+    private static function retrieveMutatorNames(array $mutatorSettings): array
+    {
+        $mutators = [];
+
+        foreach ($mutatorSettings as $mutatorOrProfile => $setting) {
+            if (array_key_exists($mutatorOrProfile, ProfileList::ALL_PROFILES)) {
+                self::registerFromProfile($mutatorOrProfile, $setting, $mutators);
+
+                continue;
+            }
+
+            if (array_key_exists($mutatorOrProfile, ProfileList::ALL_MUTATORS)) {
+                self::registerFromName($mutatorOrProfile, $setting, $mutators);
+
+                continue;
+            }
+
+            throw new InvalidArgumentException(sprintf(
+                'The profile or mutator "%s" was not recognized.',
+                $mutatorOrProfile
+            ));
+        }
+
+        return $mutators;
+    }
+
+    /**
+     * @param array<string, string>|bool           $settings
+     * @param array<string, array<string, string>> $mutators
+     */
+    private static function registerFromProfile(
+        string $profile,
+        $settings,
+        array &$mutators
```

Comment:
I'm not really a fan of passing the `$mutators` array around by reference.

But i'm not 100% sure what another solution would be.

Reply:
Perhaps, we can use `\ArrayObject` here without `&`? See https://3v4l.org/UBaZI

Reply:
Also note in case this is not clear: this array is strictly internal to the class, it is a temporary artefact. It is not what we provide as a result of `MutatorFactory::create()`